learningequality/studio
# Vueify exercise creation

## Description

Vueified exercise creation.

It doesn't contain:
- Markdown editor and related logic (preview mode, numeric values restriction for input questions) - right now there is only a simple, not working, skeleton. We decided to work on it in a separate PR (we'll be probably using a different library).
- Preventing users from accessing another exercise until a current exercise is fixed - this depends on how we'll use dialogs within the whole edit modal and also if we'll use routing or no - it will be better to work on it with edit modal high level stuff. All business logic is prepared, it just needs to be called from an appropriate place.

We discussed various UX improvements during development so the new version is not 100% clone. The most important links:
https://www.figma.com/file/NOBGTf5nCogkcQ3dEEphY6DE/Vuetify-exercise-editor?node-id=0%3A1
https://www.notion.so/learningequality/List-features-for-prioritization-bf41de54110444e88305d81f644660de
https://www.notion.so/learningequality/List-validation-cases-f002282907fd4dc089ecb4ce30e2cde4

#### Issue Addressed

https://www.notion.so/learningequality/Exercise-creation-1a085a11eb494ccab3962a22c20dd202

## Steps to Test

- [ ] *Visit http://0.0.0.0:8080/sandbox/*
- [ ] *click EDIT MODAL*
- [ ] *select "Sample exercise"*
- [ ] *Navigate to "Questions" tab*
- [ ] *Play :)! Also - I tried to examine our current version as much as possible but please have a look if I haven't forgotten some important piece.*

You can check Storybook which I use as my components development environment [here](https://pensive-wilson-0e1fa3.netlify.com) - might be helpful for components structure and interface review. You can have a look at "Notes", "Actions" (emitted events overview) and "Story" (preview of a corresponding story that is basically code that you would write to use a component, though sometimes it contains Storybook specific stuff) tabs.

## At a high level, how did you implement this?

When an assessment view visited for a selected node, a new assessment entry is created in so-called "drafts store" which is later updated as a user creates new or updates existing questions, answers, and hints. Those data are saved in a map with node ids as keys to allow users to switch between various exercise nodes without losing client data. Should be hopefully flexible enough to allow us to work with different API saving strategies.

The whole exercise editor has been decoupled from store - as discussed with @rtibbles, we might need to share it between more projects. It's still open but at least it will prevent coupling it more and more as we add updates.

## Screenshots
![Selection_013](https://user-images.githubusercontent.com/13509191/62151913-b1bfff80-b301-11e9-92a6-bf9764ee51a3.png)
![Selection_018](https://user-images.githubusercontent.com/13509191/62155872-36af1700-b30a-11e9-80a5-c4d973ceb525.png)
![Selection_014](https://user-images.githubusercontent.com/13509191/62151919-b684b380-b301-11e9-854c-a29480347c1b.png)
![Selection_015](https://user-images.githubusercontent.com/13509191/62151932-b97fa400-b301-11e9-82bd-bf66dbf4eef8.png)
![Selection_016](https://user-images.githubusercontent.com/13509191/62151943-bd132b00-b301-11e9-9733-0121effae918.png)
![Selection_017](https://user-images.githubusercontent.com/13509191/62151946-c00e1b80-b301-11e9-92bd-02e3ac9c2718.png)
